python/lsst/ts/standardscripts/maintel/closed_loop.py

```diff
@@ -0,0 +1,519 @@
+                self.exposure_time
+                + self.lsstcam.read_out_time
+                + self.lsstcam.shutter_time
+            )
+            * (2 if self.mode == ClosedLoopMode.CWFS else 1)
+            + self.lsstcam.read_out_time
+            + self.lsstcam.shutter_time
+        )
+        metadata.filter = f"{self.filter},{self.grating}"
+
+    async def take_intra_extra_focal_images(
+        self,
+    ) -> typing.Tuple[typing.Any, typing.Any]:
+        """Take intra and extra focal images.
+
+        Returns
+        -------
+        intra_image : `typing.Any`
+            Intra focal image.
+        extra_image : `typing.Any`
+            Extra focal image.
+        """
+
+        # Take intra focal image
+        self.log.debug("Moving to intra-focal position")
+
+        await self.mtcs.move_camera_hexapod(z=self.dz)
+
+        self.log.debug("Taking intra-focal image")
+
+        intra_image = await self.latiss.take_cwfs(
+            exptime=self.exposure_time,
+            n=1,
+            group_id=self.group_id,
+            filter=self.filter,
+            grating=self.grating,
+            reason="INTRA" + ("" if self.reason is None else f"_{self.reason}"),
+            program=self.program,
+        )
+
+        self.log.debug("Moving to extra-focal position")
+
+        # Hexapod offsets are relative, so need to move 2x the offset
+        # to get from the intra- to the extra-focal position.
+        # then add the offset to compensate for un-equal magnification
+        z_offset = -(self.dz * 2.0 + self.extra_focal_offset)
+        await self.mtcs.move_camera_hexapod(z=z_offset)
+
+        self.log.debug("Taking extra-focal image")
+
+        extra_image = await self.latiss.take_cwfs(
+            exptime=self.exposure_time,
+            n=1,
+            group_id=self.group_id,
+            filter=self.filter,
+            grating=self.grating,
+            reason="EXTRA" + ("" if self.reason is None else f"_{self.reason}"),
+            program=self.program,
+        )
+
+        return intra_image, extra_image
+
+    async def arun(self, checkpoint: bool = False) -> None:
+        """Perform CWFS measurements and hexapod adjustments until the
+        thresholds are reached.
+
+        Parameters
+        ----------
+        checkpoint : `bool`, optional
+            Should issue checkpoints (default=False)?
+
+        Raises
+        ------
+        RuntimeError:
+            If coordinates are malformed.
+        """
+
+        for i in range(self.max_iter):
+            self.log.debug(f"CWFS iteration {i + 1} starting...")
+
+            if checkpoint:
+                await self.checkpoint(
+                    f"[{i + 1}/{self.max_iter}]: CWFS loop starting..."
+                )
+
+                await self.checkpoint(f"[{i + 1}/{self.max_iter}]: Taking image...")
+
+            # Setting visit_id's to none so run_cwfs will take a new dataset.
+            if self.mode == ClosedLoopMode.CWFS:
```

Comment:
Instead of having each option in a if/else statement like this, I'd suggest you extract each "mode" into a function, then make a dictionary with the mode as key and the function as value. Then here you just do:

```
    await self.operation_model_handlers[self.mode]()
```

where you would define `self.operation_model_handlers` in the `__init__ method as something like:

```
self.operation_model_handlers = {
    ClosedLoopMode.CWFS: self.handle_cwfs_mode,
    ClosedLoopMode.FAM: self.handle_fam_mode,
}
```

Then `self.handle_cwfs_mode` and `self.handle_fam_mode` will have the appropriate code extracted from here.